core/src/main/java/jenkins/model/lazy/BuildReferenceMapAdapter.java

```diff
@@ -81,305 +92,227 @@ public Integer lastKey() {
 
     @Override
     public Set<Integer> keySet() {
-        return core.keySet();
+        return keySet;
     }
 
     @Override
     public Collection<R> values() {
-        return new CollectionAdapter(core.values());
+        return values;
     }
 
     @Override
     public Set<Entry<Integer, R>> entrySet() {
-        return new SetAdapter(core.entrySet());
-    }
-
-    @Override
-    public int size() {
-        return core.size();
+        return entrySet;
     }
 
     @Override
     public boolean isEmpty() {
-        return core.isEmpty();
+        return entrySet().isEmpty();
     }
 
     @Override
     public boolean containsKey(Object key) {
-        return core.containsKey(key);
+        BuildReference<R> ref = key instanceof Integer ? core.get(key) : null;
+        if (ref == null) {
+            return false;
+        }
+        // if found, resolve it and check that value is equal
+        if (!ref.isSet()) {
+            buildRefResolver.apply(ref);
+        }
+        return !ref.isUnloadable();
     }
 
-    @Override
-    public boolean containsValue(Object value) {
-        return core.containsValue(value); // TODO should this be core.containsValue(wrap(value))?
+    @SuppressWarnings("unchecked")
+    private R castValue(Object value) {
+        try {
+            return  (R) value;
+        } catch (ClassCastException e) {
+            return null;
+        }
     }
 
     @Override
-    public R get(Object key) {
-        return unwrap(core.get(key));
+    public boolean containsValue(Object value) {
+        R val = castValue(value);
+        return val != null && Objects.equals(get(buildNumberProvider.apply(val)), val);
     }
 
     @Override
-    public R put(Integer key, R value) {
-        return unwrap(core.put(key, wrap(value)));
+    public R get(Object key) {
+        return key instanceof Integer ? buildRefResolver.apply(core.get(key)) : null;
     }
 
     @Override
     public R remove(Object key) {
-        return unwrap(core.remove(key));
-    }
-
-    @Override
-    public void putAll(Map<? extends Integer, ? extends R> m) {
-        for (Entry<? extends Integer, ? extends R> e : m.entrySet())
-            put(e.getKey(), e.getValue());
-    }
-
-    @Override
-    public void clear() {
-        core.clear();
-    }
-
-    @Override
-    public boolean equals(Object o) {
-        return core.equals(o); // TODO this is wrong
-    }
-
-    @Override
-    public int hashCode() {
-        return core.hashCode();
+        R val = get(key);
+        if (val == null) {
+            return null;
+        }
+        return removeValue(val) ? val : null;
```

Comment:
Would be nicer to remove the key without resolving it, but that would violate the `remove` signature. Not sure if this is ever actually called on an unloaded build to begin with.